### DISCUSSION TITLE
Using Heyoka for propagating in LEO

### DISCUSSION THREAD

**Discussion by vishwa2710** [bluescarni/heyoka#349] [Q&A] [answered]
Hi, I recently stumbled across this, and I must say, this is really cool! I've only ever used RK based integrators, I'd never heard of taylor integrators. Thanks for the excellent documentation and explanations.
I was wondering if you had tried or have any examples of using Heyoka to propagate orbits in LEO.
Specifically, today I use Boost ODEINT to propagate satellites in LEO by leveraging Earth gravity accelerations calculated from Spherical Harmonics based models like EGM96 (from geographiclib).
Additionally we apply perturbations due to third bodies, and atmospheric drag.

I definitely to read more, but it's not immediately obvious to me how you would do that here. Happy to explain in more detail.
Again, thanks for open sourcing such a cool package. I've always been a huge fan of the work done by the Advanced Concepts Team.

**Comment by bluescarni** [comment, marked as answer]
Hi @vishwa2710 and thanks for the nice words!

heyoka is definitely suitable for propagation in LEO - in fact it powers this other ACT project about accurate & fast propagation of space debris with guaranteed collision detection:

https://github.com/esa/cascade

(note that cascade does not support the latest version of heyoka yet, I am working on a new cascade version to fix this)

In the cascade notebook examples, we are using a simplified dynamical model: only a few spherical harmonics for the Earth's figure, simplified lunisolar perturbations and a simplified nrlmsise00 model for the atmospheric drag. Other, arbitrarily complex dynamical models can also be used (see the notebooks for explanations on how to do this).

The main difficulty/drawback of heyoka (and Taylor integrators in general) is that it is not possible to plug in black-box functions to implement the right-hand side of the ODEs. That is, you cannot use an existing EGM96 model in the dynamics, rather you have to provide your own implementation of EGM96 using the heyoka expression system. This is necessary because Taylor integrators need to be able not only to evaluate the right-hand side of the ODE, but also to compute the high-order derivatives necessary to construct the Taylor series of the solution.

In heyoka, I have recently added a ``model`` submodule containing the ready-made heyoka-fied implementations of several dynamical models, including, for instance, the VSOP2013 analytical planetary theory. See here for a tutorial:

https://bluescarni.github.io/heyoka.py/notebooks/vsop2013.html

Support for several other models (JPL ephemeris, ELP2000 Lunar theory, EGM models, etc.) is planned, but it is quite a bit of work, and heyoka is a project I work on in my spare time, so it is going to take a while before all of this is available.

**Reply by vishwa2710** [reply]
Hi @bluescarni ,

Thank you so much for the very prompt response! 

> This is necessary because Taylor integrators need to be able not only to evaluate the right-hand side of the ODE, but also to compute the high-order derivatives necessary to construct the Taylor series of the solution.

Ah that makes sense, because you're using AD to calculate the derivatives for the taylor expansion.

I will definitely look into those notebooks, and the code on the VSOP2013 model.

Thanks again!

**Reply by bluescarni** [reply]
Hi @vishwa2710

Not sure if you are still interested in this, but I finally added an API for the easy formulation of Earth orbit dynamics in heyoka. You can find the documentation here:

https://bluescarni.github.io/heyoka.py/autosummary_generated/heyoka.model.eo_dynamics.html#heyoka.model.eo_dynamics

(this is for the Python bindings of heyoka but in C++ it's very similar)

This currently supports geopotential (EGM2008), third-body perturbations and atmospheric drag, with full support for up-to-date Earth orientation parameters and space weather indices.

**Reply by vishwa2710** [reply]
Hi @bluescarni !

I have been following all of your (amazing) improvements to this library. When I get some time next quarter I want to leverage it for some propagation and TLE/Orbit Determination improvements (leveraging the speed + AD based derivatives) in the astrodynamics repo I help maintain https://github.com/open-space-collective/open-space-toolkit-astrodynamics

Amazing stuff!
